core/src/main/java/com/linecorp/armeria/server/logging/structured/StructuredLogWriter.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright 2016 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server.logging.structured;
+
+import com.linecorp.armeria.common.RequestContext;
+import com.linecorp.armeria.common.logging.ResponseLog;
+
+/**
+ * An interface that a structured log writer should implement.
+ * @param <L> type of the structured log representation
+ */
+@FunctionalInterface
+public interface StructuredLogWriter<L> {
```

Comment:
`StructuredLog` is not coupled with any serialization format. It's just an intermediate representation of common log entries coupled with the procedure to extract them from given information.

Let me explain my intention with practical example(partially pseudo though). In fact I'm going to use protobuf for the production use and here's how it will look like in my current expectation:

`ProtobufKafkaSerializer.java`(this class is provided by a dependent library in our case):
```java
class ProtobufKafkaSerializer<T extends MessageLite> implements Serializer<T> {
...
    public byte[] serialize(..., T message) {
        return message.toByteArray(); // can be applied for any protobuf generated message classes
    }
}
```

Protobuf IDL will look like:
```
message ServiceFooLog {
    int64 timestamp_ms = 1;
    int64 response_time_ns = 2;
    ...
    string user_id = 10;
}
```
You should know that that this IDL will be compiled into corresponding java class named `ServiceFooLog`.

Service construction:
```java
... producer = new KakfaProducer(config, ..., new ProtobufSerializer());
service.decorate(
    KafkaStructuredLoggingService.newDecorator(
        producer, topic,
        /* StructuredLogBuilder<ServiceFooLog> */ (ctx, res) -> {
            ApacheThriftStructuredLog thriftLog = new ApacheThriftStructuredLog(ctx, res);
            return ServiceFooLog.newBuilder()
                                                .setTimestampMs(thriftLog.timestampMillis())
                                                .setResponseTimeNs(thriftLog.responseTimeNanos())
                                                ...
                                                .setUserId(extractUserIdFromHttpHeader(ctx.request())
                                                .build();
        },
        /* KeySelector<ServiceFooLog> */ (ctx, res, log) -> log.getUserId().getBytes()));
```

In order not to bring information extraction logic to serializers, letting users to build their own model within `StructuredLogBuilder` is necessary and we should have type boundary because it does not make sense to create another intermediate log representation class when there's already one which is generated from IDL and has identical fields.